python/katana/local/_graph.pyx

```diff
@@ -156,6 +188,63 @@ cdef class GraphBase:
         edge_range = self.topology().edges(n)
         return range(deref(edge_range.begin()), deref(edge_range.end()))
 
+    def get_edge_source(self, uint64_t e):
+        """
+        Get the source of an edge. This is O(log(N)) since it uses binary search to avoid having to store a source list.
+        :param e: an edge ID.
+        :return: The source node of that edge.
+        """
+        bottom = 0
+        top = self.num_nodes()
+        while bottom < top:
+            pivot = bottom + (top - bottom) // 2
+            edge_range = self.edge_ids(pivot)
+            if edge_range.start <= e < edge_range.stop:
+                return pivot
+            elif e >= edge_range.stop:
+                bottom = edge_range.stop
+            elif e < edge_range.start:
+                top = edge_range.start
+            else:
+                raise AssertionError("Unreachable")
+        raise ValueError(e)
+
+    def edges(self, n = None) -> DataFrame:
```

Comment:
Would it be possible to make the second argument a collection of nodes, and not just limit it to a single node?

Reply:
Hmm. Yes. Basically preslicing the data when it's generated. I'm not sure if there is much advantage in doing that over getting the data then slicing. The DataFrame is very cheap to make. Maybe we should remove the argument entirely. Make this a property. And then the user and project out what they want at the dataframe level.

Reply:
My thinking about slicing the data was a bit wrong. I'll make this argument accept collections and add a TODO to polish the API later.